File: app/src/main/java/com/haruhi/bismark439/haruhiism/Widgets/DayCounterProvider.java
package com.haruhi.bismark439.haruhiism.Widgets;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;

import com.haruhi.bismark439.haruhiism.R;

import static com.haruhi.bismark439.haruhiism.MainActivity.Widgets;
import static com.haruhi.bismark439.haruhiism.Widgets.ConfigurationActivity.loadWidgets;
import static com.haruhi.bismark439.haruhiism.Widgets.ConfigurationActivity.saveWidgets;
import static com.haruhi.bismark439.haruhiism.Widgets.HaruhiWidgetProvider.getDays;

/**
 * Created by Bismark439 on 28/02/2018.
 */

public class DayCounterProvider extends AppWidgetProvider {
    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        final int count = appWidgetIds.length;
        System.out.println("Widget Count: "+count);
        if(Widgets.getSize()==0){
           loadWidgets(context);
            System.out.println("Widget load: "+Widgets.getSize());
        }
        for (int i = 0; i < count; i++) {
            int widgetId = appWidgetIds[i];
            System.out.println("Widget ID: "+widgetId);
            WidgetData temp = Widgets.getByCode(widgetId);
            if(temp!=null){
                System.out.println("Found: "+widgetId);
                temp.printString();
                RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.widget_haruhi);
                long days=getDays(temp.yy,temp.mmMod,temp.dd);
                remoteViews.setTextViewText(R.id.sinceWhen, temp.name);
                if(days == 0){
                    remoteViews.setTextViewText(R.id.daysText, "Today!");
                    remoteViews.setTextViewText(R.id.days2, " ");
                }else{
                           days = Math.abs(days);
                          remoteViews.setTextViewText(R.id.daysText, days+"");
                    remoteViews.setTextViewText(R.id.days2, context.getString(R.string.days));
                }
                remoteViews.setTextColor(R.id.daysText, temp.color);
                remoteViews.setTextColor(R.id.days2, temp.color);
                remoteViews.setTextColor(R.id.sinceWhen,temp.color);
                remoteViews.setImageViewResource(R.id.widgetImage,temp.picture);
                Intent intent = new Intent(context, DayCounterProvider.class);
                intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
                intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, appWidgetIds);

                appWidgetManager.updateAppWidget(widgetId, remoteViews);
            }
        }
    }

    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        super.onDeleted(context, appWidgetIds);
        System.out.println("WIDGET DELETED called");
        if(Widgets.getSize()==0){
            loadWidgets(context);
            System.out.println("Widget load: "+Widgets.getSize());
        }
        for(int i=0;i<Widgets.getSize();i++){
            boolean found = false;
            for(int x=0;x<appWidgetIds.length;x++){
                if(Widgets.get(i).reqCode==appWidgetIds[x]){
                    found=true;
                    x=appWidgetIds.length;
                }
            }
            if(!found){
                Widgets.widgetDB.remove(i);
                if(i>0)i--;
            }
        }
        System.out.println("Modified Widget load: "+Widgets.getSize());
        saveWidgets(context);
    }
}

File: app/src/main/java/com/haruhi/bismark439/haruhiism/widgets/providers/DayCounterProvider.kt
package com.haruhi.bismark439.haruhiism.widgets.providers

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import com.haruhi.bismark439.haruhiism.Debugger
import com.haruhi.bismark439.haruhiism.model.widgetDB.WidgetDB
import com.haruhi.bismark439.haruhiism.model.widgetDB.WidgetDB.Companion.loadWidgets
import com.haruhi.bismark439.haruhiism.model.widgetDB.WidgetDao
import com.haruhi.bismark439.haruhiism.widgets.providers.WidgetCreater.onCounterClicked
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

/*

*
 * Created by Bismark439 on 28/02/2018.

*/

@DelicateCoroutinesApi
class DayCounterProvider : AppWidgetProvider() {
    companion object {
        const val ACTION_TOUCH = "com.haruhi.bismark439.haruhiism.action.TOUCH_WIDGET"
    }

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        loadWidgets(context) {
            updateUI(context, appWidgetManager, appWidgetIds)
        }
    }

    private fun updateUI(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
       // Debugger.log("Widget DB: " + WidgetDB.getSize() + " installed: " + appWidgetIds.size)
        for (widgetId in appWidgetIds) {
            val widgetData = WidgetDB.get(widgetId) ?: continue
            val ui = WidgetCreater.createFromWidgetData(context, widgetData)
            appWidgetManager.updateAppWidget(widgetId, ui)
        }
    }

    private fun updateUI(
        context: Context,
        appWidgetId: Int
    ) {
        WidgetDao.initDao(context)
        GlobalScope.launch {
            val widgetData = WidgetDao.instance.selectOnce(appWidgetId) ?: return@launch
            Debugger.log("Update on $appWidgetId called")
            val ui = WidgetCreater.createFromWidgetData(context, widgetData)
            AppWidgetManager.getInstance(context).updateAppWidget(appWidgetId, ui)
        }
    }


    override fun onDeleted(context: Context, appWidgetIds: IntArray) {
        super.onDeleted(context, appWidgetIds)
        loadWidgets(context) { removeMissingWidgets(context, appWidgetIds) }
    }

    private fun removeMissingWidgets(
        context: Context,
        appWidgetIds: IntArray
    ) {
        for (id in appWidgetIds) {
            WidgetDB.deleteWidget(context, id)
        }
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        onCounterClicked(context, intent)
        WidgetCreater.incrementView(context)
        val id = intent.getIntExtra(WidgetCreater.THIS_WIDGET_ID, -1)
        if (id != -1) {
            updateUI(context,id)
        }
    }
}


